### ISSUE TITLE
package auctex fails to be fetched

### ISSUE THREAD

**Issue by cxandru** [NixOS/nixpkgs#110796]
**Describe the bug**
At current unstable:
 - channels(root): `"nixos-21.03pre265130.f217c0ea7c1, nixpkgs-20.09pre218717.05f0934825c"`
`emacsWithPackages` fails to build auctex. The problem seems to be that it tries to fetch `https://elpa.gnu.org/packages/auctex-13.0.3.tar`, but that archive does not exist, instead for auctex the archives of all but the most up-to-date version (currently 13.0.4) are of the form `.tar.lz` (see https://elpa.gnu.org/packages/auctex.html)

**Notify maintainers**
@adisbladis I think?

**Comment by jyp** [comment]
Afaiu this bug is an instance of a general problem. Elpa does not maintain archives for older versions of packages. This means that as soon as a new version of a new package is uploaded on Elpa, it breaks nix builds, throwing away the idea of reproducible builds. This happens also when fetching org-mode, for example. Even more annoying, I cannot find any way to override the version in my config. As a workaround it seems that one can use the method described here to manually override the version: https://discourse.nixos.org/t/override-a-transitive-emacs-dependency/6700

```auctex.override

**Comment by collares** [comment]
They compress older versions for some reason. In principle, Nix could try to fetch both `X` and `X.lz` and use the first successful result.

**Comment by jorsn** [comment]
This nix function adds the archive link to an ELPA package:

```nix
lzip:
# ELPA keeps old versions only lzip-compressed.
# The current version is only available uncompressed.
# So we add the archive link to the packages.
pkg: pkg.overrideAttrs (old: {
  src = old.src.overrideAttrs (sold: {
    # The compressed archive appends a '.lz' to the url.
    # We try the archive url first, since a package is probably
    # current for a short time.
    urls = builtins.concatMap (url: ["${url}.lz" url]) sold.urls;
    downloadToTemp = true;
    postFetch = ''
      ${lzip}/bin/lzip -d $downloadedFile -c > $out || cp $downloadedFile $out
    '';
  });
})
```

The question is: Where should this go? @ttuegel @adisbladis 

1. The definition of `buildElpa`?
2. An extra fetcher `fetchElpa`?
3. Shall [emacs2nix](https:/ttuegel/emacs2nix) generate modified fetchurl args directly?
4. Anything better?

**Comment by collares** [comment]
I filed https://debbugs.gnu.org/cgi/bugreport.cgi?bug=46441 for the alternate solution of offering lzma files (as well as uncompressed files) even for the current version of packages.

**Comment by jorsn** [comment]
> I filed https://debbugs.gnu.org/cgi/bugreport.cgi?bug=46441

According to the bug discussion we could use the `externals/<package>` branches in <http://git.savannah.gnu.org/cgit/emacs/elpa.git>` to build the elpa packages.

We somehow need to get the git revisions. There are at least two possibilities for doing this which copy (A) or use directly (B) the code used by ELPA to build the packages. Both start with a checked out elpa `main` branch.

First, we clone and prepare the repo of 254 MB:

```sh
sh $ git clone https://git.savannah.gnu.org/git/emacs/elpa.git
sh $ cd elpa
sh $ make # checks out branch elpa-admin to worktree admin/ and links Makefile
```

We store some helper functions in `rev.el`:

```elisp
(defun main-file (pkgname) (elpaa--main-file (elpaa--get-package-spec pkgname)))

(defun release-rev (pkgname)
  (elpaa--get-release-revision
    (expand-file-name pkgname "packages")
    (elpaa--get-package-spec pkgname)))

(defun with-first-arg (f) (apply f (cons (pop command-line-args-left) nil)))

(defun cli-princ-main-file () (princ (with-first-arg 'main-file)))

(defun cli-princ-release-rev () (princ (with-first-arg 'release-rev)))
```

## A

```sh
sh $ pkg=auctex
sh $ main_file="$(emacs --batch -Q -l admin/elpa-admin.el -l rev.el -f cli-princ-main-file)"
sh $ git log -n1 --oneline --no-patch --pretty=format:%H \
        -L "/^;;* *\(Package-\)\?Version:/,+1:$main_file" origin/externals/"$pkg"
```

## B

This uses the `git log` command line from *A* internally, only for a checked out directory `packages/<package>`.

```sh
sh $ pkg=auctex
sh $ make packages/"$pkg"
sh $ emacs --batch -Q -l admin/elpa-admin.el -l rev.el -f cli-princ-release-rev "$pkg"
```

**Comment by collares** [comment]
This was fixed in #132937.
